File: internal/gitengine/wire.go
package gitengine

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

// DefaultMaxFrame bounds every helper allocation to 16 MiB. Large hunks must
// be rejected explicitly or split by a future protocol version.
const DefaultMaxFrame uint32 = 16 << 20

// FrameTag is a four-byte wire frame discriminator.
type FrameTag [4]byte

var (
	TagHello      = FrameTag{'H', 'E', 'L', 'O'}
	TagBatchBegin = FrameTag{'B', 'G', 'I', 'N'}
	TagCommit     = FrameTag{'C', 'M', 'I', 'T'}
	TagFileBegin  = FrameTag{'F', 'B', 'E', 'G'}
	TagHunk       = FrameTag{'H', 'U', 'N', 'K'}
	TagHunkBegin  = FrameTag{'H', 'B', 'G', 'N'}
	TagHunkAdd    = FrameTag{'H', 'A', 'D', 'D'}
	TagHunkEnd    = FrameTag{'H', 'E', 'N', 'D'}
	TagFileEnd    = FrameTag{'F', 'E', 'N', 'D'}
	TagCommitEnd  = FrameTag{'C', 'E', 'N', 'D'}
	TagBatchEnd   = FrameTag{'B', 'E', 'N', 'D'}
	TagError      = FrameTag{'E', 'R', 'R', 'O'}
	TagQuit       = FrameTag{'Q', 'U', 'I', 'T'}
)

// Frame is the bounded transport unit. Payload always owns its bytes.
type Frame struct {
	Tag     FrameTag
	BatchID uint64
	Payload []byte
}

// WriteFrame writes one length-prefixed frame.
func WriteFrame(w io.Writer, frame Frame, max uint32) error {
	if max == 0 {
		max = DefaultMaxFrame
	}
	length := uint64(12) + uint64(len(frame.Payload))
	if length > uint64(max) {
		return fmt.Errorf("frame length %d exceeds limit %d", length, max)
	}
	var header [16]byte
	binary.BigEndian.PutUint32(header[:4], uint32(length))
	copy(header[4:8], frame.Tag[:])
	binary.BigEndian.PutUint64(header[8:16], frame.BatchID)
	if err := writeAll(w, header[:]); err != nil {
		return err
	}
	if len(frame.Payload) == 0 {
		return nil
	}
	return writeAll(w, frame.Payload)
}

func writeAll(w io.Writer, data []byte) error {
	for len(data) > 0 {
		n, err := w.Write(data)
		if err != nil {
			return err
		}
		if n == 0 {
			return io.ErrShortWrite
		}
		data = data[n:]
	}
	return nil
}

// ReadFrame reads one frame and rejects truncated or oversized input before
// allocating the payload.
func ReadFrame(r io.Reader, max uint32) (Frame, error) {
	if max == 0 {
		max = DefaultMaxFrame
	}
	var prefix [4]byte
	if _, err := io.ReadFull(r, prefix[:]); err != nil {
		return Frame{}, err
	}
	length := binary.BigEndian.Uint32(prefix[:])
	if length < 12 {
		return Frame{}, fmt.Errorf("frame length %d is shorter than header", length)
	}
	if length > max {
		return Frame{}, fmt.Errorf("frame length %d exceeds limit %d", length, max)
	}
	data := make([]byte, length)
	if _, err := io.ReadFull(r, data); err != nil {
		return Frame{}, err
	}
	var tag FrameTag
	copy(tag[:], data[:4])
	return Frame{
		Tag:     tag,
		BatchID: binary.BigEndian.Uint64(data[4:12]),
		Payload: data[12:],
	}, nil
}

// HelloFrame encodes negotiated capabilities.
func HelloFrame(capabilities Capabilities) Frame {
	var payload []byte
	payload = binary.BigEndian.AppendUint16(payload, capabilities.ProtocolVersion)
	payload = binary.BigEndian.AppendUint16(payload, capabilities.OIDLength)
	payload = appendBool(payload, capabilities.AllStatuses)
	payload = appendBool(payload, capabilities.FindCopies)
	return Frame{Tag: TagHello, Payload: payload}
}

// DecodeHello decodes and validates a HELO frame.
func DecodeHello(frame Frame) (Capabilities, error) {
	if frame.Tag != TagHello || frame.BatchID != 0 || len(frame.Payload) != 6 {
		return Capabilities{}, errors.New("invalid HELO frame")
	}
	capabilities := Capabilities{
		ProtocolVersion: binary.BigEndian.Uint16(frame.Payload[:2]),
		OIDLength:       binary.BigEndian.Uint16(frame.Payload[2:4]),
		AllStatuses:     frame.Payload[4] != 0,
		FindCopies:      frame.Payload[5] != 0,
	}
	if frame.Payload[4] > 1 || frame.Payload[5] > 1 {
		return Capabilities{}, errors.New("invalid HELO boolean")
	}
	if err := capabilities.Validate(); err != nil {
		return Capabilities{}, err
	}
	return capabilities, nil
}

// BatchBeginFrame encodes a bounded request.
func BatchBeginFrame(request BatchRequest) (Frame, error) {
	if err := request.Validate(0); err != nil {
		return Frame{}, err
	}
	if uint64(len(request.Commits)) > uint64(^uint32(0)) {
		return Frame{}, errors.New("batch commit count exceeds protocol limit")
	}
	var payload []byte
	payload = binary.BigEndian.AppendUint32(payload, uint32(len(request.Commits)))
	for _, oid := range request.Commits {
		if err := validateOID(oid, false); err != nil {
			return Frame{}, err
		}
		payload = appendBytes(payload, oid)
	}
	return Frame{Tag: TagBatchBegin, BatchID: request.ID, Payload: payload}, nil
}

// DecodeBatchBegin decodes a BGIN request and owns every OID.
func DecodeBatchBegin(frame Frame) (BatchRequest, error) {
	if frame.Tag != TagBatchBegin || frame.BatchID == 0 {
		return BatchRequest{}, errors.New("invalid BGIN frame")
	}
	d := decoder{data: frame.Payload}
	count, err := d.u32()
	if err != nil {
		return BatchRequest{}, err
	}
	if uint64(count)*24 > uint64(len(d.data)-d.pos) {
		return BatchRequest{}, errors.New("batch commit count exceeds bounded payload")
	}
	commits := make([]OID, 0, count)
	for range count {
		oid, err := d.bytes()
		if err != nil {
			return BatchRequest{}, err
		}
		if err := validateOID(oid, false); err != nil {
			return BatchRequest{}, err
		}
		commits = append(commits, OID(oid))
	}
	if err := d.done(); err != nil {
		return BatchRequest{}, err
	}
	request := BatchRequest{ID: frame.BatchID, Commits: commits}
	if err := request.Validate(0); err != nil {
		return BatchRequest{}, err
	}
	return request, nil
}

// RecordFrame encodes a typed record using its lifecycle frame tag.
func RecordFrame(batchID uint64, record Record) (Frame, error) {
	if batchID == 0 {
		return Frame{}, errors.New("record batch ID is zero")
	}
	payload, err := encodeRecordPayload(record)
	if err != nil {
		return Frame{}, err
	}
	var tag FrameTag
	switch record.Kind {
	case RecordCommit:
		tag = TagCommit
	case RecordFile:
		tag = TagFileBegin
	case RecordHunk:
		tag = TagHunk
	}
	return Frame{Tag: tag, BatchID: batchID, Payload: payload}, nil
}

// DecodeRecord decodes CMIT, FBEG, or HUNK and owns every byte sequence.
func DecodeRecord(frame Frame) (Record, error) {
	if frame.BatchID == 0 {
		return Record{}, errors.New("record frame batch ID is zero")
	}
	var kind RecordKind
	switch frame.Tag {
	case TagCommit:
		kind = RecordCommit
	case TagFileBegin:
		kind = RecordFile
	case TagHunk:
		kind = RecordHunk
	default:
		return Record{}, fmt.Errorf("frame %.4s is not a record", frame.Tag)
	}
	record, err := decodeRecordPayload(kind, frame.Payload)
	if err != nil {
		return Record{}, err
	}
	if err := validateRecord(record); err != nil {
		return Record{}, err
	}
	return record, nil
}

// HunkFrames returns one HUNK frame when it fits, otherwise a bounded
// HBGN/HADD+/HEND continuation sequence. The assembled record remains subject
// to the receiver's independent MaxRecord bound.
func HunkFrames(batchID uint64, hunk HunkRecord, max uint32) ([]Frame, error) {
	if batchID == 0 {
		return nil, errors.New("hunk batch ID is zero")
	}
	if max == 0 {
		max = DefaultMaxFrame
	}
	if max <= 12 {
		return nil, errors.New("frame limit cannot hold payload")
	}
	record := Record{Kind: RecordHunk, Hunk: hunk}
	if err := validateRecord(record); err != nil {
		return nil, err
	}
	regularSize := uint64(12 + 4 + len(hunk.Commit) + 4 + len(hunk.NewPath) + 8 + 4 + len(hunk.Added) + 1)
	if regularSize <= uint64(max) {
		frame, err := RecordFrame(batchID, record)
		if err != nil {
			return nil, err
		}
		return []Frame{frame}, nil
	}
	beginPayload := encodeHunkBegin(hunk)
	if uint64(12+len(beginPayload)) > uint64(max) {
		return nil, errors.New("HBGN metadata exceeds frame limit")
	}
	if len(hunk.Added) == 0 {
		return nil, errors.New("continued hunk requires nonempty added bytes")
	}
	frames := []Frame{{Tag: TagHunkBegin, BatchID: batchID, Payload: beginPayload}}
	chunkSize := int(max) - 12
	for offset := 0; offset < len(hunk.Added); offset += chunkSize {
		end := min(offset+chunkSize, len(hunk.Added))
		frames = append(frames, Frame{Tag: TagHunkAdd, BatchID: batchID, Payload: bytes.Clone(hunk.Added[offset:end])})
	}
	frames = append(frames, Frame{Tag: TagHunkEnd, BatchID: batchID, Payload: appendBool(nil, hunk.MissingFinalNewline)})
	return frames, nil
}

// DecodeHunkBegin decodes continuation metadata without added bytes.
func DecodeHunkBegin(frame Frame) (HunkRecord, error) {
	if frame.Tag != TagHunkBegin || frame.BatchID == 0 {
		return HunkRecord{}, errors.New("invalid HBGN frame")
	}
	d := decoder{data: frame.Payload}
	commit, err := d.oid(false)
	if err != nil {
		return HunkRecord{}, err
	}
	path, err := d.bytes()
	if err != nil {
		return HunkRecord{}, err
	}
	position, err := d.u64()
	if err != nil {
		return HunkRecord{}, err
	}
	if err := d.done(); err != nil {
		return HunkRecord{}, err
	}
	hunk := HunkRecord{Commit: commit, NewPath: path, NewPosition: position}
	if err := validateRecord(Record{Kind: RecordHunk, Hunk: hunk}); err != nil {
		return HunkRecord{}, err
	}
	return hunk, nil
}

// DecodeHunkEnd decodes the final-newline state discovered after streamed
// added bytes have already been emitted.
func DecodeHunkEnd(frame Frame) (bool, error) {
	if frame.Tag != TagHunkEnd || frame.BatchID == 0 || len(frame.Payload) != 1 {
		return false, errors.New("invalid HEND frame")
	}
	if frame.Payload[0] > 1 {
		return false, errors.New("invalid HEND boolean")
	}
	return frame.Payload[0] == 1, nil
}

func encodeHunkBegin(hunk HunkRecord) []byte {
	var payload []byte
	payload = appendBytes(payload, hunk.Commit)
	payload = appendBytes(payload, hunk.NewPath)
	payload = binary.BigEndian.AppendUint64(payload, hunk.NewPosition)
	return payload
}

// BatchEndFrame encodes the helper-confirmed result.
func BatchEndFrame(result BatchResult) Frame {
	payload := []byte{byte(result.Status)}
	payload = binary.BigEndian.AppendUint64(payload, result.Commits)
	payload = binary.BigEndian.AppendUint64(payload, result.Files)
	payload = binary.BigEndian.AppendUint64(payload, result.Hunks)
	return Frame{Tag: TagBatchEnd, BatchID: result.ID, Payload: payload}
}

// DecodeBatchEnd decodes a BEND frame.
func DecodeBatchEnd(frame Frame) (BatchResult, error) {
	if frame.Tag != TagBatchEnd || frame.BatchID == 0 || len(frame.Payload) != 25 {
		return BatchResult{}, errors.New("invalid BEND frame")
	}
	status := CompletionStatus(frame.Payload[0])
	if status != CompletionOK && status != CompletionRejected {
		return BatchResult{}, errors.New("invalid BEND status")
	}
	return BatchResult{
		ID:      frame.BatchID,
		Status:  status,
		Commits: binary.BigEndian.Uint64(frame.Payload[1:9]),
		Files:   binary.BigEndian.Uint64(frame.Payload[9:17]),
		Hunks:   binary.BigEndian.Uint64(frame.Payload[17:25]),
	}, nil
}

// ErrorFrame encodes a typed terminal helper error.
func ErrorFrame(batchID uint64, kind ErrorKind, op string, err error) Frame {
	payload := []byte{byte(kind)}
	payload = appendBytes(payload, []byte(op))
	payload = appendBytes(payload, []byte(err.Error()))
	return Frame{Tag: TagError, BatchID: batchID, Payload: payload}
}

// DecodeError decodes an ERRO frame.
func DecodeError(frame Frame) (*EngineError, error) {
	if frame.Tag != TagError || len(frame.Payload) < 1 {
		return nil, errors.New("invalid ERRO frame")
	}
	kind := ErrorKind(frame.Payload[0])
	if kind < ErrorProtocol || kind > ErrorUnsupported {
		return nil, errors.New("invalid ERRO kind")
	}
	d := decoder{data: frame.Payload[1:]}
	op, err := d.bytes()
	if err != nil {
		return nil, err
	}
	message, err := d.bytes()
	if err != nil {
		return nil, err
	}
	if err := d.done(); err != nil {
		return nil, err
	}
	return &EngineError{Kind: kind, BatchID: frame.BatchID, Op: string(op), Err: errors.New(string(message))}, nil
}

func encodeRecordPayload(record Record) ([]byte, error) {
	canonical, err := CanonicalRecord(record)
	if err != nil {
		return nil, err
	}
	return canonical[1:], nil
}

func decodeRecordPayload(kind RecordKind, payload []byte) (Record, error) {
	d := decoder{data: payload}
	record := Record{Kind: kind}
	var err error
	switch kind {
	case RecordCommit:
		r := &record.Commit
		if r.OID, err = d.oid(false); err != nil {
			return Record{}, err
		}
		if r.Message, err = d.bytes(); err != nil {
			return Record{}, err
		}
		if r.AuthorName, err = d.bytes(); err != nil {
			return Record{}, err
		}
		if r.AuthorEmail, err = d.bytes(); err != nil {
			return Record{}, err
		}
		v, e := d.u64()
		if e != nil {
			return Record{}, e
		}
		r.AuthorUnixSeconds = int64(v)
		u, e := d.u32()
		if e != nil {
			return Record{}, e
		}
		r.AuthorUTCOffsetMinutes = int32(u)
		if r.HasAuthor, err = d.boolean(); err != nil {
			return Record{}, err
		}
		if r.HasAuthorTime, err = d.boolean(); err != nil {
			return Record{}, err
		}
	case RecordFile:
		r := &record.File
		if r.Commit, err = d.oid(false); err != nil {
			return Record{}, err
		}
		status, e := d.byte()
		if e != nil {
			return Record{}, e
		}
		r.Status = Status(status)
		if r.OldMode, err = d.u32(); err != nil {
			return Record{}, err
		}
		if r.NewMode, err = d.u32(); err != nil {
			return Record{}, err
		}
		if r.OldOID, err = d.oid(true); err != nil {
			return Record{}, err
		}
		if r.NewOID, err = d.oid(true); err != nil {
			return Record{}, err
		}
		if r.OldPath, err = d.bytes(); err != nil {
			return Record{}, err
		}
		if r.NewPath, err = d.bytes(); err != nil {
			return Record{}, err
		}
		if r.Binary, err = d.boolean(); err != nil {
			return Record{}, err
		}
	case RecordHunk:
		r := &record.Hunk
		if r.Commit, err = d.oid(false); err != nil {
			return Record{}, err
		}
		if r.NewPath, err = d.bytes(); err != nil {
			return Record{}, err
		}
		if r.NewPosition, err = d.u64(); err != nil {
			return Record{}, err
		}
		if r.Added, err = d.bytes(); err != nil {
			return Record{}, err
		}
		if r.MissingFinalNewline, err = d.boolean(); err != nil {
			return Record{}, err
		}
	}
	if err := d.done(); err != nil {
		return Record{}, err
	}
	return record, nil
}

func validateOID(oid []byte, allowEmpty bool) error {
	if allowEmpty && len(oid) == 0 {
		return nil
	}
	if len(oid) != 20 && len(oid) != 32 {
		return fmt.Errorf("object ID length %d", len(oid))
	}
	return nil
}

type decoder struct {
	data []byte
	pos  int
}

func (d *decoder) byte() (byte, error) {
	if d.pos >= len(d.data) {
		return 0, io.ErrUnexpectedEOF
	}
	v := d.data[d.pos]
	d.pos++
	return v, nil
}

func (d *decoder) boolean() (bool, error) {
	v, err := d.byte()
	if err != nil {
		return false, err
	}
	if v > 1 {
		return false, errors.New("invalid boolean")
	}
	return v == 1, nil
}

func (d *decoder) u32() (uint32, error) {
	if len(d.data)-d.pos < 4 {
		return 0, io.ErrUnexpectedEOF
	}
	v := binary.BigEndian.Uint32(d.data[d.pos : d.pos+4])
	d.pos += 4
	return v, nil
}

func (d *decoder) u64() (uint64, error) {
	if len(d.data)-d.pos < 8 {
		return 0, io.ErrUnexpectedEOF
	}
	v := binary.BigEndian.Uint64(d.data[d.pos : d.pos+8])
	d.pos += 8
	return v, nil
}

func (d *decoder) bytes() ([]byte, error) {
	n, err := d.u32()
	if err != nil {
		return nil, err
	}
	if uint64(n) > uint64(len(d.data)-d.pos) {
		return nil, io.ErrUnexpectedEOF
	}
	v := append([]byte(nil), d.data[d.pos:d.pos+int(n)]...)
	d.pos += int(n)
	return v, nil
}

func (d *decoder) oid(allowEmpty bool) (OID, error) {
	v, err := d.bytes()
	if err != nil {
		return nil, err
	}
	if err := validateOID(v, allowEmpty); err != nil {
		return nil, err
	}
	return OID(v), nil
}

func (d *decoder) done() error {
	if d.pos != len(d.data) {
		return fmt.Errorf("%d trailing payload bytes", len(d.data)-d.pos)
	}
	return nil
}

File: contrib/gitengine/gix-helper/src/wire.rs
use std::collections::HashSet;
use std::io::{self, Read, Write};

pub(crate) const VERSION: u16 = 1;
pub(crate) const MAX_FRAME: u32 = 16 << 20;
pub(crate) const MAX_RECORD: usize = 256 << 20;

pub(crate) const HELO: [u8; 4] = *b"HELO";
pub(crate) const BGIN: [u8; 4] = *b"BGIN";
pub(crate) const CMIT: [u8; 4] = *b"CMIT";
pub(crate) const FBEG: [u8; 4] = *b"FBEG";
pub(crate) const HUNK: [u8; 4] = *b"HUNK";
pub(crate) const HBGN: [u8; 4] = *b"HBGN";
pub(crate) const HADD: [u8; 4] = *b"HADD";
pub(crate) const HEND: [u8; 4] = *b"HEND";
pub(crate) const FEND: [u8; 4] = *b"FEND";
pub(crate) const CEND: [u8; 4] = *b"CEND";
pub(crate) const BEND: [u8; 4] = *b"BEND";
pub(crate) const ERRO: [u8; 4] = *b"ERRO";
pub(crate) const QUIT: [u8; 4] = *b"QUIT";

#[derive(Debug, Eq, PartialEq)]
pub(crate) struct Frame {
    pub tag: [u8; 4],
    pub batch_id: u64,
    pub payload: Vec<u8>,
}

pub(crate) fn read_frame(input: &mut impl Read) -> io::Result<Frame> {
    let mut prefix = [0; 4];
    input.read_exact(&mut prefix)?;
    let length = u32::from_be_bytes(prefix);
    if length < 12 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "frame shorter than header",
        ));
    }
    if length > MAX_FRAME {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "frame exceeds 16 MiB limit",
        ));
    }
    let mut header = [0; 12];
    input.read_exact(&mut header)?;
    let mut tag = [0; 4];
    tag.copy_from_slice(&header[..4]);
    let batch_id = u64::from_be_bytes(header[4..].try_into().expect("fixed-width batch ID"));
    let payload_len = usize::try_from(length - 12).expect("u32 fits usize");
    let mut payload = vec![0; payload_len];
    input.read_exact(&mut payload)?;
    Ok(Frame {
        tag,
        batch_id,
        payload,
    })
}

pub(crate) fn write_frame(
    output: &mut impl Write,
    tag: [u8; 4],
    batch_id: u64,
    payload: &[u8],
) -> io::Result<()> {
    let length = 12usize
        .checked_add(payload.len())
        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "frame length overflow"))?;
    if length > usize::try_from(MAX_FRAME).expect("u32 fits usize") {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "frame exceeds 16 MiB limit",
        ));
    }
    output.write_all(
        &u32::try_from(length)
            .expect("bounded frame length")
            .to_be_bytes(),
    )?;
    output.write_all(&tag)?;
    output.write_all(&batch_id.to_be_bytes())?;
    output.write_all(payload)
}

pub(crate) fn hello_payload(oid_len: usize) -> Vec<u8> {
    let mut payload = Vec::with_capacity(6);
    payload.extend_from_slice(&VERSION.to_be_bytes());
    payload.extend_from_slice(
        &u16::try_from(oid_len)
            .expect("OID width fits u16")
            .to_be_bytes(),
    );
    payload.extend_from_slice(&[0, 0]);
    payload
}

pub(crate) fn decode_batch(frame: &Frame, oid_len: usize) -> io::Result<Vec<Vec<u8>>> {
    if frame.tag != BGIN || frame.batch_id == 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "invalid BGIN frame",
        ));
    }
    let mut decoder = Decoder::new(&frame.payload);
    let count = usize::try_from(decoder.u32()?).expect("u32 fits usize");
    let minimum = count
        .checked_mul(4 + oid_len)
        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "commit count overflow"))?;
    if minimum > decoder.remaining() {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "commit count exceeds payload",
        ));
    }
    let mut seen = HashSet::with_capacity(count);
    let mut commits = Vec::with_capacity(count);
    for _ in 0..count {
        let oid = decoder.bytes()?.to_vec();
        if oid.len() != oid_len {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "OID width differs from HELO",
            ));
        }
        if !seen.insert(oid.clone()) {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "duplicate OID in batch",
            ));
        }
        commits.push(oid);
    }
    decoder.finish()?;
    Ok(commits)
}

pub(crate) fn append_bytes(payload: &mut Vec<u8>, bytes: &[u8]) -> io::Result<()> {
    let length = u32::try_from(bytes.len()).map_err(|_| {
        io::Error::new(
            io::ErrorKind::InvalidInput,
            "byte string exceeds protocol limit",
        )
    })?;
    payload.extend_from_slice(&length.to_be_bytes());
    payload.extend_from_slice(bytes);
    Ok(())
}

pub(crate) fn write_hunk(
    output: &mut impl Write,
    batch_id: u64,
    commit: &[u8],
    path: &[u8],
    new_position: u64,
    added: &[u8],
    missing_final_newline: bool,
) -> io::Result<()> {
    if added.len() > MAX_RECORD {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "hunk exceeds 256 MiB limit",
        ));
    }
    let mut payload = Vec::with_capacity(
        4 + commit.len() + 4 + path.len() + 8 + 4 + added.len().min(64 << 10) + 1,
    );
    append_bytes(&mut payload, commit)?;
    append_bytes(&mut payload, path)?;
    payload.extend_from_slice(&new_position.to_be_bytes());
    let fixed_len = payload.len();
    let regular_len = 12usize
        .checked_add(fixed_len)
        .and_then(|n| n.checked_add(4 + added.len() + 1))
        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "hunk length overflow"))?;
    if regular_len <= usize::try_from(MAX_FRAME).expect("u32 fits usize") {
        append_bytes(&mut payload, added)?;
        payload.push(u8::from(missing_final_newline));
        return write_frame(output, HUNK, batch_id, &payload);
    }
    if 12 + payload.len() > usize::try_from(MAX_FRAME).expect("u32 fits usize") {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "HBGN metadata exceeds frame limit",
        ));
    }
    write_frame(output, HBGN, batch_id, &payload)?;
    let chunk_size = usize::try_from(MAX_FRAME - 12).expect("u32 fits usize");
    for chunk in added.chunks(chunk_size) {
        write_frame(output, HADD, batch_id, chunk)?;
    }
    write_frame(output, HEND, batch_id, &[u8::from(missing_final_newline)])
}

pub(crate) fn error_payload(kind: u8, operation: &str, message: &str) -> Vec<u8> {
    let mut payload = vec![kind];
    // Error strings originate locally and are bounded well below a frame.
    let _ = append_bytes(&mut payload, operation.as_bytes());
    let _ = append_bytes(&mut payload, message.as_bytes());
    payload
}

pub(crate) struct Decoder<'a> {
    input: &'a [u8],
    position: usize,
}

impl<'a> Decoder<'a> {
    pub(crate) fn new(input: &'a [u8]) -> Self {
        Self { input, position: 0 }
    }

    fn remaining(&self) -> usize {
        self.input.len() - self.position
    }

    pub(crate) fn u32(&mut self) -> io::Result<u32> {
        if self.remaining() < 4 {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                "truncated u32",
            ));
        }
        let value = u32::from_be_bytes(
            self.input[self.position..self.position + 4]
                .try_into()
                .expect("checked width"),
        );
        self.position += 4;
        Ok(value)
    }

    pub(crate) fn bytes(&mut self) -> io::Result<&'a [u8]> {
        let length = usize::try_from(self.u32()?).expect("u32 fits usize");
        if length > self.remaining() {
            return Err(io::Error::new(
                io::ErrorKind::UnexpectedEof,
                "truncated bytes",
            ));
        }
        let bytes = &self.input[self.position..self.position + length];
        self.position += length;
        Ok(bytes)
    }

    pub(crate) fn finish(self) -> io::Result<()> {
        if self.position == self.input.len() {
            Ok(())
        } else {
            Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "trailing payload bytes",
            ))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn frame_bytes(tag: [u8; 4], batch: u64, payload: &[u8]) -> Vec<u8> {
        let mut out = Vec::new();
        write_frame(&mut out, tag, batch, payload).unwrap();
        out
    }

    #[test]
    fn frame_round_trip_preserves_arbitrary_bytes() {
        let payload = b"\0\xff\nraw";
        let frame = read_frame(&mut frame_bytes(CMIT, 9, payload).as_slice()).unwrap();
        assert_eq!(
            frame,
            Frame {
                tag: CMIT,
                batch_id: 9,
                payload: payload.to_vec()
            }
        );
    }

    #[test]
    fn rejects_short_truncated_and_oversized_frames() {
        assert!(read_frame(&mut [0, 0, 0, 11].as_slice()).is_err());
        assert!(read_frame(&mut [0, 0, 0, 12].as_slice()).is_err());
        let too_large = (MAX_FRAME + 1).to_be_bytes();
        assert!(read_frame(&mut too_large.as_slice()).is_err());
    }

    #[test]
    fn batch_supports_both_hash_widths_and_rejects_duplicates() {
        for width in [20, 32] {
            let oid = vec![7; width];
            let mut payload = 1u32.to_be_bytes().to_vec();
            append_bytes(&mut payload, &oid).unwrap();
            let got = decode_batch(
                &Frame {
                    tag: BGIN,
                    batch_id: 1,
                    payload,
                },
                width,
            )
            .unwrap();
            assert_eq!(got, vec![oid]);
        }
        let oid = vec![1; 20];
        let mut payload = 2u32.to_be_bytes().to_vec();
        append_bytes(&mut payload, &oid).unwrap();
        append_bytes(&mut payload, &oid).unwrap();
        assert!(
            decode_batch(
                &Frame {
                    tag: BGIN,
                    batch_id: 1,
                    payload
                },
                20
            )
            .is_err()
        );
    }

    #[test]
    fn streams_hunks_larger_than_sixteen_mib() {
        let added = vec![b'x'; (17 << 20) + 3];
        let mut out = Vec::new();
        write_hunk(&mut out, 4, &[1; 20], b"p", 8, &added, true).unwrap();
        let mut input = out.as_slice();
        let begin = read_frame(&mut input).unwrap();
        assert_eq!(begin.tag, HBGN);
        let mut reconstructed = Vec::new();
        loop {
            let frame = read_frame(&mut input).unwrap();
            match frame.tag {
                HADD => reconstructed.extend_from_slice(&frame.payload),
                HEND => {
                    assert_eq!(frame.payload, [1]);
                    break;
                }
                tag => panic!("unexpected tag {tag:?}"),
            }
        }
        assert_eq!(reconstructed, added);
        assert!(input.is_empty());
    }
}
